src/ATen/function_wrapper.py

```diff
@@ -219,10 +219,23 @@ def is_mutable_formal_argument(argument, option):
     return argument.get('output') or option['inplace'] and argument['name'] == 'self'
 
 
-def to_return_type(arg, option):
+# Note [Clang and "Attempted to construct a reference element in a tuple with an rvalue"]
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# to_return_type can be used in contexts where we are doing multiple
+# return, in which case it will be incorporated into an expression like:
+# std::tuple<at::Tensor, at::Tensor>.  Originally, this code assumed it
+# would be valid to return a reference, e.g., as in
+# std::tuple<at::Tensor, const at::Tensor&>.  This is accepted by GCC
+# and early versions of Clang, but not by recent Clang; c.f.
+# https://bugs.llvm.org/show_bug.cgi?id=20855   To work around this
+# problem, we don't create references in this case.
+#
+# TODO: Figure out how to portably return references in multiple return, as
+# it *is* more efficient (avoids refcount.)
+def to_return_type(arg, option, multiple_return):
```

Comment:
Something funky is going on here. The only time & should be used is when we are returning an input tensor as an output, and this bug does not describe that case. Can you attach the generated code with the clang error before this change? I am wondering if some change introduce a bug in this pathway that wasn't caught.

Reply:
Here is one example bad generated code:

```
std::tuple<Tensor,const Tensor &> VariableType::orgqr(const Tensor & self, const Tensor & input2) const {
    auto& self_ = unpack(self, "self", 0);
    auto& input2_ = unpack(input2, "input2", 1);
    auto flags = Function::flags({ self, input2 });
    auto grad_fn = std::make_shared<OrgqrBackward>();
    auto ret = as_variable(baseType->orgqr(self_, input2_));
    wrap_output(ret, std::move(flags), grad_fn);
    if (jit::tracer::isTracing({ self, input2 })) {
      jit::Node *n = jit::tracer::recordTrace( "orgqr", { self, input2 }, ret );
      (void)n;
    }
    return std::tuple<Tensor,const Tensor &>(std::move(ret));
}
```

Reply:
This whole function seems suspicious to me because it is not an inplace or an _out function, but it is still returning an input. 

We should

1. keep & returns for when the return is an _out or a inplace
2. change the functions that sneekily return an input as output to return Tensor rather than Tensor &. Because a client will not know anyway that it is happening. How many of these are there?

```
std::tuple<Tensor,const Tensor &> CPUFloatType::ormqr(const Tensor & self, const Tensor & input2, const Tensor & input3) {
    auto result_ = new CPUFloatTensor(context);
    auto result = Tensor(result_,false);
    auto self_ = checked_cast<CPUFloatTensor>(self.pImpl,"self",1, false);
    auto input2_ = checked_cast<CPUFloatTensor>(input2.pImpl,"input2",2, false);
    auto input3_ = checked_cast<CPUFloatTensor>(input3.pImpl,"input3",3, false);
    THFloatTensor_ormqr(result_->tensor, self_->tensor, input2_->tensor, input3_->tensor, "L", "N");
    bool maybe_scalar = self.dim() == 0 && input2.dim() == 0 && input3.dim() == 0;
    result_->maybeScalar(maybe_scalar);
    self_->maybeScalar(maybe_scalar);
    return std::tuple<Tensor, const Tensor &>(result, self);
}
```

Reply:
It looks like it's just orgqr and ormqr.

Reply:
The Declarations.cwrap for orgqr and ormqr should be modified so that it just returns argument 0 (and not argument 1).

The confusion likely came from the fact that the LAPACK implementations operate in-place, but this isn't an issue because the TH code first clones the input before calling the LAPACK function.

Reply:
@colesbury tells me that it's a typo in the definitions. New fix was force pushed.